aws-logs-querydefinitions/src/main/java/software/amazon/logs/querydefinitions/ReadHandler.java

```diff
@@ -0,0 +1,55 @@
+package software.amazon.logs.querydefinitions;
+
+import software.amazon.awssdk.services.cloudwatchlogs.model.DescribeQueryDefinitionsRequest;
+import software.amazon.awssdk.services.cloudwatchlogs.model.DescribeQueryDefinitionsResponse;
+import software.amazon.awssdk.services.cloudwatchlogs.model.InvalidParameterException;
+import software.amazon.awssdk.services.cloudwatchlogs.model.QueryDefinition;
+import software.amazon.awssdk.services.cloudwatchlogs.model.ServiceUnavailableException;
+import software.amazon.cloudformation.exceptions.CfnInvalidRequestException;
+import software.amazon.cloudformation.exceptions.CfnNotFoundException;
+import software.amazon.cloudformation.exceptions.CfnServiceInternalErrorException;
+import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
+import software.amazon.cloudformation.proxy.Logger;
+import software.amazon.cloudformation.proxy.ProgressEvent;
+import software.amazon.cloudformation.proxy.OperationStatus;
+import software.amazon.cloudformation.proxy.ResourceHandlerRequest;
+
+import java.util.Optional;
+
+public class ReadHandler extends BaseHandler<CallbackContext> {
+
+    @Override
+    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
+        final AmazonWebServicesClientProxy proxy,
+        final ResourceHandlerRequest<ResourceModel> request,
+        final CallbackContext callbackContext,
+        final Logger logger) {
+
+        final ResourceModel model = request.getDesiredResourceState();
+
+        DescribeQueryDefinitionsResponse describeQueryDefinitionsResponse;
+        String nextToken = null;
+        do {
+            try {
+                describeQueryDefinitionsResponse = proxy.injectCredentialsAndInvokeV2(Translator.translateToReadRequest(model, nextToken),
+                        ClientBuilder.getLogsClient()::describeQueryDefinitions);
+                nextToken = describeQueryDefinitionsResponse.nextToken();
+                Optional<QueryDefinition> queryDefinition = describeQueryDefinitionsResponse.queryDefinitions().stream()
+                        .filter(qd -> qd.queryDefinitionId().equals(model.getQueryDefinitionId()))
```

Comment:
Where does validation happen? Is it guaranteed that `model.getQueryDefinitionId()` is set or valid? I'm guessing CFN should be validating against the JSON Schema before calling the function, but I'm wondering what about the condition when there is a bug in CFN? Should we have more validation and logging up front?

Reply:
yes for these calls as the ID is set as a primary identifier. I added a validation step in the create handler to ensure that we are not calling creates with querydef set as I have the as readOnly, but besides that I think it is best practice to let the CFN WFs handle the validation (afaict)